client/state/sites/vouchers/reducer.js

```diff
@@ -28,6 +33,28 @@ import { itemsSchema } from './schema';
 export const items = ( state = {}, action ) => {
 	const { siteId } = action;
 	switch ( action.type ) {
+		case SITE_VOUCHERS_ASSIGN:
+			switch ( action.serviceType ) {
+				case GOOGLE_AD_CREDITS:
+					const currentVouchers = state[ siteId ]
+						? ( state[ siteId ][ GOOGLE_AD_CREDITS ] || [] )
+						: [];
+
+					const vouchers = currentVouchers.splice( 0 );
+					vouchers.push( action.voucher );
+
+					return Object.assign(
+						{},
+						state,
+						{
+							[ siteId ]: Object.assign(
+								{},
+								{ [ GOOGLE_AD_CREDITS ]: vouchers }
+							)
+						}
+					);
+			}
```

Comment:
These return types look complicated enough that the work we are doing to return them immutably seems to be quite large. It could be a good chance to use immutability helpers such as [immutable.js](https://facebook.github.io/immutable-js/docs/#/) or the [React immutability helpers](http://facebook.github.io/react/docs/update.html).

The code here ends up cloning everything. Granted, the data structures should be small and performance probably won't be significant enough to matter, but I find it hard to figure out by reading the code what exactly is supposed to be returned.

The way to do this with **Immutable.js** would look something like this:

```js
case SITE_VOUCHERS_ASSIGN:
	return state.updateIn( [ siteId, serviceType ], vouchers => vouchers.concat( action.voucher ) );
```

The React immutability helpers keep things as Javascript objects while Immutable.js results in a different data-structure that needs to be converted or used in the view code. The benefit to both of these is that an update to one site and one service type will leave the rest of the data structure intact, only cloning the small pieces that changed.